### ISSUE TITLE
Track approximation activity is not visible or stuck

### ISSUE THREAD

**Issue by sonora** [osmandapp/OsmAnd#15025]
I get this dialog with no indication of activity or progress, progress indicator on it seems broken?

![Screenshot_20220806-184457_OsmAnd~](https://user-images.githubusercontent.com/845735/183258696-e75e8463-4245-47ba-adbe-7c5b40f04119.jpg)

Also: I have the impression that changing the routing profile by tapping on it stalls the calculation process entirely, or at least enormously delays any result?

**Comment by sonora** [comment]
Another observation iindicating that the profile switching may not be clean:
- Have the map in pedestrian view
- tap a GPX track, on its content menu hit the nav button
- When the route appears, tap the new "Attach to the road"  button
- Wait for the pedestrian route to finish calculation and hit "Apply"
- Now go back the the main Navigation dialog: On it the "Use elevation data" button for the ped mode has disappeared. Need to restart the app to get it back...

![Screenshot_20220806-191701_OsmAnd~](https://user-images.githubusercontent.com/845735/183259253-87b8629a-c9a3-499c-89dd-fb4d42a72dae.jpg)

**Comment by vshcherb** [comment]
Indication is on the top left and it could be more prominent of course.

**Comment by sonora** [comment]
> Indication is on the top left and it could be more prominent of course.

Yes, but even that seems "stuck" there at 5%  in these particular use cases, then eventually jumps to 100% (if ever).

**Comment by ivanPyrohivskyi** [comment]
2022-10-06_16-36_Thu.gpx

**Comment by Zirochkabila** [comment]
OsmAnd~ 4.3.0#13892mqta, released: 2022-10-22
The loading progress bar is blinking + in album orientation doesn't fit on the screen

https://user-images.githubusercontent.com/104760013/197346279-d240d488-9d62-43ed-8247-6767f318b016.mp4

**Comment by sonora** [comment]
I find no improvement, I guess this was only fixed fod OpenGL and needs fixing in core-legacy also?

**Comment by sonora** [comment]
Looks like it is an initialization issue: With the latest build, if I wait long enough for the first approximation (~2 minutes), subsequent navigation mode changes or approximations work fast then.

**Comment by sonora** [comment]
Second part of this issue (as described in https://github.com/osmandapp/OsmAnd/issues/15025#issuecomment-1207250103 above) is still not fixed. User needs to restart the app to get the 'Use elevatio data' option back of ped mode is used for the GPX routing. Similarly for bicycle mode.

**Comment by sonora** [comment]
Upon playing with this in recent days, I con only re-emphasize that this is a 100% reproducible initialization issue: The first time I use "Attach to roads" in every app session, it takes like 3 minutes wait on the "Threshold distance" screen before the progress indicator even starts to move.

Once I have completed an "Attach to roads", all subsequent 'attaches' start instantaneously and finish fast (even if I select another app mode than originally). Until the next fresh start of OsmAnd.

Can anyone reproduce this? I am tempted to re-open this, unless I am the only one experiencing this?

**Comment by vshcherb** [comment]
@Zirochkabila can you reproduce?

**Comment by sonora** [comment]
_Some issues don't come up on clean test devices, only on devices used in the real world...  ;)_

I have further researched this: The decisive point impacting this is **how many maps (obf files) you have in your osmand-folder.** I usually have about 70. When I go down to just 5, the issue disappears, and the "Attach to roads" is essentially instantaneous.

- [ ] So we need to research what happens during the initial '10 second wait' documented in the video above, and why is scales so badly with increasing number of maps. There must be some init step, probably obsolete, but eating time and likely CPU cycles.

Somehow I feel like this is similar to #6740, where we had suddenly introduced such a wait time for "regular" routing...

In a nutshell again how to reproduce:
- (Build with no OpenGL libs, render engine setting set to Version 1)
- Perform a fresh app start. Issue occurs only for the first "Attach o roads" after a fresh app start (no matter how long ago this was), all subsequent tries are fast.
- Select a track you display on the map by tapping on it (e.g. a track previously recorded by OsmAnd)
- On its context menu, hit the Navigation-arrow button
- When the route config dialog comes up, scroll up and hit "Attach to roads"
- On the interim info screen hit "Continue"
- Next screen hangs for several minutes if you have enough maps on your device, then it suddenly finishes ok.
- All subsequent tries of the same procedure are fast, just like you show.
- Moving maps to backup diminishes the issue.

**Comment by sonora** [comment]
In addition to the finding above, here's another hint: When I use 
- "Menu" - "Plan a route"
- "Create a route"
- tap a few arbitrary points
- then tap "Options", "Route between the points", seleect e.g. "Walking",

the conversion to a route starts immediately and progressses swiftly, and you can on the map visibly follow the progress of the points being moved to the paths.

If on the other hand instead of "Route between the pounts" I  use "Options " - "Attach to roads" there, I see nothing happening at all? A bug, I guess?

I am currently not clear why for "Attach to roads" we seem to apply a different logic than for "Route between the points", shouldn't the algorithm to use for both actions be the same?

**Comment by sonora** [comment]
Please note the new finding above that **the bug is only in the native code**! Reopening...

**Comment by vshcherb** [comment]
I can't observe the bug or at least it's not visible in my use cases. Let's see if @Zirochkabila will find something similar

**Comment by sonora** [comment]
I think it's e.g. visible in his recording `telegram-cloud-document-2-5436257462233604336.mp4 `above from 0:00 to 0:12.

**Comment by Zirochkabila** [comment]
OsmAnd~ 4.3.0#13945mqta, released: 2022-11-14
86 maps - 10.1 GB - **Not Reproduced**

Indeed, when you attach to the roads when you first turn on the application, there is a delay of 12 seconds, but not for minutes - you can see the progress gradually.
All subsequent attempts are much faster


https://user-images.githubusercontent.com/104760013/201652859-a6af6469-e299-424d-bc2f-f84a962737a9.mp4

**Comment by sonora** [comment]
As I said,
- **(1)** it is 12 seconds on your device, and it **scales with the number of ofb files you have in OsmAnd's root data folder:** When I reduce my files to 5, the issue disappears, but I normally have about 70 files there, and it is then **3 minutes...**
- **(2)** More importantly: **This is a bug in Native code only!** Activating the Develoment plugins's Setting "Safe GPX approximation" never shows the issue at all, no matter how many obfs you have in your folder.

**Comment by vshcherb** [comment]
I can confirm that there is a delay ( to me 30% of the whole time) but it happens on first calculation longer, I'll check if it's easy to fix or it will be for next release

**Comment by sonora** [comment]
@vshcherb Confirmed working - good job!  ;)
